### ISSUE TITLE
Error: Cannot find module 'typescript'

### ISSUE THREAD

**Issue by raucao** [readthedocs/readthedocs.org#7591]
(Sorry if this is the wrong repo for this particular question/problem. Feel free to move it wherever you want.)

We finally managed to set up [working builds with TypeDoc](https://readthedocs.org/projects/remotestoragejs/builds/) and the new SpinxJS recently. But when I started pushing new code today, the builds started failing, without us having changed anything else about the setup.

TypeDoc now complains that it can't import the `typescript` module anymore:

> Error: Cannot find module 'typescript'

(We're using a shell-out in `conf.py` to `npm install` the modules, and the `typescript` module is installed as a dependency in that build directory.)

Any idea how we could fix this perhaps? Maybe TypeScript could be installed as a global npm module for the build images?

**Comment by raucao** [comment]
Thanks, could be. But it runs just fine with the stable version of node.js outside of RTD.

Maybe it would be possible to provide the current stable node.js version (12) on RTD, instead of version 8? I already tried to fix this by rolling back dependencies, but we can't roll back everything to outdated versions with a few thousand security issues in the tree.

**Comment by stsewd** [comment]
We install the default version that is included in ubuntu (18.04 currently)

**Comment by humitos** [comment]
> Maybe it would be possible to provide the current stable node.js version (12) on RTD, instead of version 8?

You could use `conda` for your environment and install the version of node that you need with it. Another option is use a hack similar to https://github.com/readthedocs/readthedocs-docker-images/issues/107

**Comment by raucao** [comment]
> You could use `conda` for your environment and install the version of node that you need with it.

That sounds interesting. Could you give me a hint for how I would do this in a RTD build? Unfortunately, I have zero knowledge about Conda and very limited knowledge about the RTD environment.

> Another option is use a hack similar to [readthedocs/readthedocs-docker-images#107](https://github.com/readthedocs/readthedocs-docker-images/issues/107)

It looks like that would only update npm, because it's installing npm as an npm module. But node.js would still be stuck two major versions behind LTS.

**Comment by lissyx** [comment]
We started hitting the same a few days ago as well on our deepspeech doc builds. Quickly investigating, a `npm list` within `/usr/local/lib` does show weird things (full log at https://readthedocs.org/api/v2/build/12252844.txt):
```
/usr/local/lib
├─┬ jsdoc@3.6.6
│ ├── @babel/parser@7.12.3
│ ├── bluebird@3.7.2
│ ├─┬ catharsis@0.8.11
│ │ └── lodash@4.17.20
│ ├── escape-string-regexp@2.0.0
│ ├─┬ js2xmlparser@4.0.1
│ │ └── xmlcreate@2.0.3
│ ├─┬ klaw@3.0.0
│ │ └── graceful-fs@4.2.4
│ ├─┬ markdown-it@10.0.0
│ │ ├─┬ argparse@1.0.10
│ │ │ └── sprintf-js@1.0.3
│ │ ├── entities@2.0.3
│ │ ├── linkify-it@2.2.0
│ │ ├── mdurl@1.0.1
│ │ └── uc.micro@1.0.6
│ ├── markdown-it-anchor@5.3.0
│ ├── marked@0.8.2
│ ├── mkdirp@1.0.4
│ ├── requizzle@0.2.3
│ ├── strip-json-comments@3.1.1
│ ├── taffydb@2.6.2
│ └── underscore@1.10.2
├─┬ typedoc@0.19.2
│ ├─┬ fs-extra@9.0.1
│ │ ├── at-least-node@1.0.0
│ │ ├── graceful-fs@4.2.4
│ │ ├── jsonfile@6.0.1
│ │ └── universalify@1.0.0
│ ├─┬ handlebars@4.7.6
│ │ ├── minimist@1.2.5
│ │ ├── neo-async@2.6.2
│ │ ├── source-map@0.6.1
│ │ ├── uglify-js@3.11.3
│ │ └── wordwrap@1.0.0
│ ├── highlight.js@10.3.1
│ ├── lodash@4.17.20
│ ├── lunr@2.3.9
│ ├── marked@1.2.0
│ ├─┬ minimatch@3.0.4
│ │ └─┬ brace-expansion@1.1.11
│ │   ├── balanced-match@1.0.0
│ │   └── concat-map@0.0.1
│ ├── progress@2.0.3
│ ├── semver@7.3.2
│ ├─┬ shelljs@0.8.4
│ │ ├─┬ glob@7.1.6
│ │ │ ├── fs.realpath@1.0.0
│ │ │ ├─┬ inflight@1.0.6
│ │ │ │ └── wrappy@1.0.2
│ │ │ ├── inherits@2.0.4
│ │ │ ├── once@1.4.0
│ │ │ └── path-is-absolute@1.0.1
│ │ ├── interpret@1.4.0
│ │ └─┬ rechoir@0.6.2
│ │   └─┬ resolve@1.17.0
│ │     └── path-parse@1.0.6
│ └── typedoc-default-themes@0.11.4
└── UNMET PEER DEPENDENCY typescript@3.9.x || 4.0.x

npm ERR! peer dep missing: typescript@3.9.x || 4.0.x, required by typedoc@0.19.2
```

**Comment by raucao** [comment]
Just FYI: I also tried a global install of the typescript package in our config, before the build is being run, but it didn't help.

**Comment by lissyx** [comment]
> We started hitting the same a few days ago as well on our deepspeech doc builds.

Looks like ensuring `typedoc` installed by our local `npm install` call to be first in `PATH` was enough to unblock.

**Comment by raucao** [comment]
@lissyx Mind sharing your config perhaps?

**Comment by lissyx** [comment]
> @lissyx Mind sharing your config perhaps?

https://github.com/mozilla/DeepSpeech/commit/19eeadd0f3770a47c5faa1075bde0243b37e98ff

**Comment by raucao** [comment]
The TypeScript missing error is gone when setting the PATH, but now I'm stuck with node.js 8 being too old to understand `catch` without an error argument, which is used in `fs-extra` by `typedoc`. Apparently the only fix is to upgrade to node.js 10.0 at least. :/

https://readthedocs.org/projects/remotestoragejs/builds/12253674/

(node.js 10 is from April 2018, and the current stable version is two major versions newer than 10.)

**Comment by stsewd** [comment]
I'm closing this in favor of https://github.com/readthedocs/readthedocs-docker-images/issues/107, please subscribe to that issue for updates.
